src/monattack.cpp

```diff
@@ -2255,6 +2255,171 @@ void mattack::tentacle(monster *z, int index)
+    }
+    else {
+        zz->setpos( pt );
+    }
+        range--;
+        if( seen ) {
+            g->draw();
+        }
+    }
+    if( seen ) {
+        add_msg( _("The %s's arms fly out and pull %s back!"), z->name().c_str(), target->disp_name().c_str() );
+    }
+}
+
+void mattack::grab(monster *z, int index)
+{
+    if( !z->can_act() ) {
+        return;
+    }
+    Creature *target = z->attack_target();
+    if( target == nullptr || rl_dist( z->pos(), target->pos() ) > 1 ) {
+        return;
+    }
+
+    player *foe = dynamic_cast< player* >( target );
+
+    z->reset_special(index); // Reset timer
+    z->moves -= 80;
+    bool uncanny = target->uncanny_dodge();
+    auto msg_type = foe == &g->u ? m_warning : m_info;
+    if( uncanny || dodge_check(z, target) ){
+        if( foe != nullptr ) {
+                target->add_msg_player_or_npc( msg_type, _("The %s gropes at you, but you dodge!"),
+                                                      _("The %s gropes at <npcname>, but they dodge!"),
+                                            z->name().c_str() );
+
+    } else {target->add_msg_player_or_npc( msg_type, _("The %s gropes at you, but misses!"),
+                                            _("The %s gropes at <npcname>, but it misses!"),
+                                z->name().c_str() );
+    }
+    if( !uncanny ) {
+        target->on_dodge( z, z->type->melee_skill * 2 );
+    }
+    return;
+    }
+    target->add_msg_player_or_npc(m_bad, _("%s grabs you!"), _("%s grabs <npcname>!"),
+                                z->disp_name().c_str());
+
+    if ( target->has_grab_break_tec() && target->get_grab_resist() > 0 && target->get_dex() > target->get_str() ?
+        dice(target->get_dex(), 10) : dice(target->get_str(), 10) > dice(8, 10)) {
+        target->add_msg_player_or_npc(m_good, _("You break the grab!"),
+                                    _("<npcname> breaks the grab!"));
+        return;
+        }
+    target->add_effect("grabbed", 2);
+}
+
+void mattack::grab_pull(monster *z, int index)
+{
+    if( !z->can_act() ) {
+        return;
+    }
+    Creature *target = z->attack_target();
+    monster *zz = dynamic_cast<monster*>(target);
+    if( target == nullptr || rl_dist( z->pos(), target->pos() ) > 1 ) {
+        return;
+    }
+
+    player *foe = dynamic_cast< player* >( target );
+
+    grab(z, index);
+
+    if (target->has_effect("grabbed")){
+        tripoint target_square = z->pos() - (target->pos() - z->pos());
+        if (z->can_move_to(target_square) && target->stability_check(z, 1) ){
+            tripoint zpt = z->pos();
+            z->move_to(target_square);
+            if (foe != nullptr){
+                if( target->is_player() && ( zpt.x < SEEX * int(MAPSIZE / 2) || zpt.y < SEEY * int(MAPSIZE / 2) ||
+            zpt.x >= SEEX * (1 + int(MAPSIZE / 2)) || zpt.y >= SEEY * (1 + int(MAPSIZE / 2)) ) ) {
+            g->update_map( zpt.x, zpt.y );
+        }
+        if (foe->in_vehicle) {
```

Comment:
Null foe alert.

When in doubt, always test the zeds by pitting them against other zeds rather than the player. Only test the player if the zeds vs zeds work.